src/lib/y2storage/proposal/phys_vol_calculator.rb

```diff
@@ -0,0 +1,213 @@
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "y2storage/disk_size"
+require "byebug"
+
+module Y2Storage
+  class Proposal
+    # Class used by SpaceDistribution to find the best distribution of LVM
+    # physical volumes.
+    class PhysVolCalculator
+      # Initialize.
+      #
+      # @param all_spaces [Array<FreeDiskSpace>] Disk spaces that could
+      #     potentially contain physical volumes
+      # @param lvm_helper [Proposal::LvmHelper] contains information about the
+      #     LVM planned volumes and how to make space for them
+      def initialize(all_spaces, lvm_helper)
+        @all_spaces = all_spaces
+        @lvm_helper = lvm_helper
+      end
+
+      # Extended distribution that includes a planned volume for every
+      # physical volumes that would be necessary to fulfill the LVM requirements
+      #
+      # When using LVM, the number of possible distributions increases a lot.
+      # For every space in the disk we can decide to place an LVM PV or not.
+      # This method explores all the options and finds the best one.
+      #
+      # @param distribution [Proposal::SpaceDistribution] initial distribution
+      #     containing only the non-LVM volumes
+      #
+      # @return [Proposal::SpaceDistribution, nil] nil if it's impossible to
+      #     allocate all the needed physical volumes
+      def add_physical_volumes(distribution)
+        best = nil
+        all_spaces.permutation.each do |sorted_spaces|
+          candidate = processed(distribution, sorted_spaces)
+          next unless candidate
+          best = candidate if !best || candidate.better_than(best) < 0
+        end
+        best
+      end
+
+    protected
+
+      attr_reader :lvm_helper, :all_spaces
+
+      # Returns a new SpaceDistribution created by assigning a physical volume
+      # to each space, following the given order, until the goal is reached.
+      #
+      # Returns nil if it's not possible to create a distribution of physical
+      # volumes that guarantees the requirements set by lvm_helper.
+      #
+      # @param distribution [Proposal::SpaceDistribution] initial distribution
+      #     containing only the non-LVM volumes
+      # @param sorted_spaces [Array<FreeDiskSpace>]
+      #
+      # @return [Proposal::SpaceDistribution, nil]
+      def processed(distribution, sorted_spaces)
+        volumes = {}
+        missing_size = lvm_helper.missing_space
+        result = nil
+
+        sorted_spaces.each do |space|
+          available_size = estimated_available_size(space, distribution)
+          next unless available_size > lvm_helper.min_pv_size
```

Comment:
`>=`  as it seems that `#min_pv_size` includes at least `#extent_size` usable data?
Or even `if ... <  ...`?